File: examples/async_pipeline.py
#!/usr/bin/env python3
# SPDX-FileCopyrightText: Copyright 2025 Au-Zone Technologies
# SPDX-License-Identifier: Apache-2.0
"""
Async pipeline example for ARA-2 NPU with circular buffer.

Demonstrates pipelined inference using a ring of model instances, each
with its own DMA-BUF tensors. While the NPU executes on one set of
buffers, the CPU fills the next set and reads results from the previous.

Pipeline stages::

    ┌──────────┐   ┌──────────┐   ┌──────────┐
    │ Fill (CPU)│──▶│Infer(NPU)│──▶│Read (CPU)│
    └──────────┘   └──────────┘   └──────────┘
         slot[i]      slot[i-1]      slot[i-2]

Usage::

    python async_pipeline.py /path/to/model.dvm [iterations] [depth]

Requirements:
    edgefirst-ara2
    ARA-2 proxy service running (dvproxy — systemd unit: ara2.service or dvproxy.service)
"""

from __future__ import annotations

import argparse
import time


def fill_inputs(model, frame: int) -> None:
    """Simulate CPU preprocessing by writing a pattern into input tensors."""
    import numpy as np

    val = frame % 256
    for i in range(model.n_inputs):
        shape = model.input_shape(i)
        data = np.full((shape[0], shape[1], shape[2]), val, dtype=np.uint8)
        model.set_input_tensor(i, data)


def read_outputs(model) -> int:
    """Simulate CPU postprocessing by reading the first byte of each output."""
    checksum = 0
    for i in range(model.n_outputs):
        data = model.get_output_tensor(i)
        checksum = (checksum + int(data.flat[0])) & 0xFF
    return checksum


def main() -> None:
    parser = argparse.ArgumentParser(description="ARA-2 async pipeline example")
    parser.add_argument("model", help="Path to a compiled .dvm model file")
    parser.add_argument(
        "iterations", nargs="?", type=int, default=100, help="Number of iterations (> 0)"
    )
    parser.add_argument(
        "depth", nargs="?", type=int, default=2, help="Ring buffer depth (> 0)"
    )
    args = parser.parse_args()
    if args.iterations <= 0:
        parser.error("iterations must be > 0")
    if args.depth <= 0:
        parser.error("depth must be > 0")

    import edgefirst_ara2 as ara2

    print(f"edgefirst-ara2 v{ara2.__version__}")
    print()

    # ── Connect ──────────────────────────────────────────────────────
    session = ara2.Session.create_via_unix_socket(ara2.DEFAULT_SOCKET)
    print(f"Connected via {session.socket_type} socket")

    endpoints = session.list_endpoints()
    assert len(endpoints) > 0, "No endpoints available"
    endpoint = endpoints[0]

    # ── Create circular buffer of model slots ────────────────────────
    n = args.iterations
    depth = args.depth
    slots = []
    for _ in range(depth):
        model = endpoint.load_model(args.model)
        model.allocate_tensors()
        slots.append(model)

    print(
        f"Pipeline: {depth} slots x ({slots[0].n_inputs} inputs, "
        f"{slots[0].n_outputs} outputs) with DMA-BUF tensors"
    )

    # Warm-up each slot
    for slot in slots:
        slot.run()
    print("Warm-up complete\n")

    # ── Synchronous baseline ─────────────────────────────────────────
    print(f"=== Synchronous baseline (fill → run → read) x {n} ===")
    start = time.perf_counter()
    for frame in range(n):
        slot = slots[0]
        fill_inputs(slot, frame)
        slot.run()
        read_outputs(slot)
    sync_ms = (time.perf_counter() - start) * 1000
    sync_fps = n / (sync_ms / 1000)
    print(f"  Total: {sync_ms:.2f}ms, Avg: {sync_ms / n:.2f}ms/iter, Throughput: {sync_fps:.1f} fps")

    # ── Pipelined async ──────────────────────────────────────────────
    print(f"\n=== Pipelined async (depth={depth}) x {n} ===")
    start = time.perf_counter()

    inflight: list[object | None] = [None] * depth

    for frame in range(n + depth):
        slot_idx = frame % depth

        # Wait for previous request on this slot (if any).
        if inflight[slot_idx] is not None:
            inflight[slot_idx].wait()
            read_outputs(slots[slot_idx])
            inflight[slot_idx] = None

        # Fill and submit if there are still frames to process.
        if frame < n:
            fill_inputs(slots[slot_idx], frame)
            inflight[slot_idx] = slots[slot_idx].submit()

    pipeline_ms = (time.perf_counter() - start) * 1000
    pipeline_fps = n / (pipeline_ms / 1000)
    print(
        f"  Total: {pipeline_ms:.2f}ms, Avg: {pipeline_ms / n:.2f}ms/iter, "
        f"Throughput: {pipeline_fps:.1f} fps"
    )

    # ── Summary ──────────────────────────────────────────────────────
    speedup = sync_ms / pipeline_ms
    print(f"\nSpeedup: {speedup:.2f}x (sync {sync_ms:.2f}ms vs pipeline {pipeline_ms:.2f}ms)")
    print(f"In-flight requests remaining: {session.inflight_count()}")
    print("\nDone.")


if __name__ == "__main__":
    main()

File: examples/async_pipeline.rs
// SPDX-FileCopyrightText: Copyright 2025 Au-Zone Technologies
// SPDX-License-Identifier: Apache-2.0

//! # Async Pipeline Example
//!
//! Demonstrates pipelined inference with a circular buffer of DMA-BUF
//! tensor sets. Each "slot" in the ring is a separate model instance with
//! its own registered DMA-BUF tensors. While the NPU executes slot N, the
//! CPU fills slot N+1 (and reads results from slot N-1), achieving full
//! CPU/NPU overlap.
//!
//! ## Pipeline Stages
//!
//! ```text
//!   ┌──────────┐   ┌──────────┐   ┌──────────┐
//!   │ Fill (CPU)│──▶│Infer(NPU)│──▶│Read (CPU)│
//!   └──────────┘   └──────────┘   └──────────┘
//!        slot[i]      slot[i-1]      slot[i-2]
//! ```
//!
//! ## Usage
//!
//! ```text
//! async_pipeline /path/to/model.dvm [iterations] [depth]
//! ```
//!
//! - `iterations`: total inferences to run (default: 100)
//! - `depth`: ring buffer depth / number of model slots (default: 2)

use ara2::{DEFAULT_TIMEOUT_MS, InferRequest, Model, Session};
use edgefirst_hal::tensor::TensorTrait;
use std::{env, path::PathBuf, process, time::Instant};

/// Simulate CPU preprocessing by writing a pattern into all input tensors.
fn fill_inputs(model: &mut Model, frame: usize) {
    let val = (frame % 256) as u8;
    for i in 0..model.n_inputs() {
        let tensor = model.input_tensor(i);
        let mut map = tensor.map().expect("Failed to map input tensor");
        map.fill(val);
    }
}

/// Simulate CPU postprocessing by reading the first byte of each output.
fn read_outputs(model: &Model) -> u8 {
    let mut checksum: u8 = 0;
    for i in 0..model.n_outputs() {
        let tensor = model.output_tensor(i);
        let map = tensor.map().expect("Failed to map output tensor");
        checksum = checksum.wrapping_add(map[0]);
    }
    checksum
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        eprintln!("Usage: {} <model.dvm> [iterations] [depth]", args[0]);
        process::exit(1);
    }
    let model_path = PathBuf::from(&args[1]);
    let iterations: usize = args.get(2).and_then(|s| s.parse().ok()).unwrap_or(100);
    if iterations == 0 {
        eprintln!("Error: iterations must be > 0");
        process::exit(1);
    }
    let depth: usize = args.get(3).and_then(|s| s.parse().ok()).unwrap_or(2);
    if depth == 0 {
        eprintln!("Error: depth must be > 0");
        process::exit(1);
    }

    // ── Connect ─────────────────────────────────────────────────────
    let session = Session::create_via_unix_socket(ara2::DEFAULT_SOCKET)
        .expect("Failed to connect to ARA-2 proxy");
    println!("Connected to ARA-2 proxy");

    let endpoints = session.list_endpoints().expect("Failed to list endpoints");
    assert!(!endpoints.is_empty(), "No endpoints available");
    let endpoint = &endpoints[0];

    // ── Create circular buffer of model slots ───────────────────────
    // Each slot is a separate Model instance with its own DMA-BUF
    // tensors. This allows the NPU to execute on one set of buffers
    // while the CPU fills/reads another.
    let mut slots: Vec<Model> = Vec::with_capacity(depth);
    for _ in 0..depth {
        let mut model = endpoint
            .load_model_from_file(&model_path)
            .expect("Failed to load model");
        model
            .allocate_tensors(None)
            .expect("Failed to allocate tensors");
        slots.push(model);
    }

    println!(
        "Pipeline: {} slots x ({} inputs, {} outputs) with DMA-BUF tensors",
        depth,
        slots[0].n_inputs(),
        slots[0].n_outputs()
    );

    // Warm-up each slot
    for slot in &mut slots {
        slot.run().expect("Warm-up failed");
    }
    println!("Warm-up complete\n");

    // ── Baseline: synchronous ───────────────────────────────────────
    println!("=== Synchronous baseline (fill → run → read) x {iterations} ===");
    let start = Instant::now();
    for frame in 0..iterations {
        let slot = &mut slots[0];
        fill_inputs(slot, frame);
        slot.run().expect("Sync inference failed");
        let _ = read_outputs(slot);
    }
    let sync_elapsed = start.elapsed();
    println!(
        "  Total: {:.2}ms, Avg: {:.2}ms/iter, Throughput: {:.1} fps",
        sync_elapsed.as_secs_f64() * 1000.0,
        sync_elapsed.as_secs_f64() * 1000.0 / iterations as f64,
        iterations as f64 / sync_elapsed.as_secs_f64(),
    );

    // ── Pipelined: circular buffer ──────────────────────────────────
    println!("\n=== Pipelined async (depth={depth}) x {iterations} ===");
    let start = Instant::now();

    // In-flight request ring: one Option<InferRequest> per slot.
    let mut inflight: Vec<Option<InferRequest>> = (0..depth).map(|_| None).collect();

    for frame in 0..iterations + depth {
        let slot_idx = frame % depth;

        // Wait for the previous request on this slot (if any).
        if let Some(req) = inflight[slot_idx].take() {
            req.wait(DEFAULT_TIMEOUT_MS).expect("Wait failed");
            // Read outputs from the slot that just completed.
            let _ = read_outputs(&slots[slot_idx]);
        }

        // If there are still frames to submit, fill and submit.
        if frame < iterations {
            fill_inputs(&mut slots[slot_idx], frame);
            let req = slots[slot_idx].submit().expect("Submit failed");
            inflight[slot_idx] = Some(req);
        }
    }

    let pipeline_elapsed = start.elapsed();
    println!(
        "  Total: {:.2}ms, Avg: {:.2}ms/iter, Throughput: {:.1} fps",
        pipeline_elapsed.as_secs_f64() * 1000.0,
        pipeline_elapsed.as_secs_f64() * 1000.0 / iterations as f64,
        iterations as f64 / pipeline_elapsed.as_secs_f64(),
    );

    // ── Summary ─────────────────────────────────────────────────────
    let speedup = sync_elapsed.as_secs_f64() / pipeline_elapsed.as_secs_f64();
    println!(
        "\nSpeedup: {:.2}x (sync {:.2}ms vs pipeline {:.2}ms)",
        speedup,
        sync_elapsed.as_secs_f64() * 1000.0,
        pipeline_elapsed.as_secs_f64() * 1000.0,
    );

    let count = session
        .inflight_count()
        .expect("Failed to get inflight count");
    println!("In-flight requests remaining: {count}");
}
